pkg/registry/core/service/storage/rest.go

```diff
@@ -541,33 +586,383 @@ func (r *REST) ConvertToTable(ctx context.Context, object runtime.Object, tableO
 	return r.services.ConvertToTable(ctx, object, tableOptions)
 }
 
-// When allocating we always use BySpec, when releasing we always use ByClusterIP
-func (r *REST) getAllocatorByClusterIP(service *api.Service) ipallocator.Interface {
-	if !utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) || r.secondaryServiceIPs == nil {
-		return r.serviceIPs
+func (rs *REST) allocClusterIPs(service *api.Service, toAlloc map[api.IPFamily]string) (map[api.IPFamily]string, error) {
+	allocated := make(map[api.IPFamily]string)
+
+	for family, ip := range toAlloc {
+		allocator := rs.serviceIPAllocatorsByFamily[family] // should always be there, as we pre validate
+		if ip == "" {
+			allocatedIP, err := allocator.AllocateNext()
+			if err != nil {
+				return allocated, errors.NewInternalError(fmt.Errorf("failed to allocate a serviceIP: %v", err))
+			}
+			allocated[family] = allocatedIP.String()
+		} else {
+			parsedIP := net.ParseIP(ip)
+			if err := allocator.Allocate(parsedIP); err != nil {
+				// TODO: when validation becomes versioned, this gets more complicated.
+				el := field.ErrorList{field.Invalid(field.NewPath("spec", "clusterIPs"), service.Spec.ClusterIPs, fmt.Sprintf("failed to allocated ip:%v with error:%v", ip, err))}
+				return allocated, errors.NewInvalid(api.Kind("Service"), service.Name, el)
+			}
+			allocated[family] = ip
+		}
+	}
+	return allocated, nil
+}
+
+// releases clusterIPs per family
+func (rs *REST) releaseClusterIPs(toRelease map[api.IPFamily]string) (map[api.IPFamily]string, error) {
+	if toRelease == nil {
+		return nil, nil
 	}
 
-	secondaryAllocatorCIDR := r.secondaryServiceIPs.CIDR()
-	if netutil.IsIPv6String(service.Spec.ClusterIP) == netutil.IsIPv6CIDR(&secondaryAllocatorCIDR) {
-		return r.secondaryServiceIPs
+	released := make(map[api.IPFamily]string)
+	for family, ip := range toRelease {
+		allocator, ok := rs.serviceIPAllocatorsByFamily[family]
+		if !ok {
+			// cluster was configured for dual stack, then single stack
+			klog.V(4).Infof("delete service. Not releasing ClusterIP:%v because IPFamily:%v is no longer configured on server", ip, family)
+			continue
+		}
+
+		parsedIP := net.ParseIP(ip)
+		if err := allocator.Release(parsedIP); err != nil {
+			return released, err
+		}
+		released[family] = ip
 	}
 
-	return r.serviceIPs
+	return released, nil
 }
 
-func (r *REST) getAllocatorBySpec(service *api.Service) ipallocator.Interface {
-	if !utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) ||
-		service.Spec.IPFamily == nil ||
-		r.secondaryServiceIPs == nil {
-		return r.serviceIPs
+// standard allocator for dualstackgate==Off, hard wired dependency
+// and ignores policy, families and clusterIPs
+func (rs *REST) allocServiceClusterIP(service *api.Service) (map[api.IPFamily]string, error) {
```

Comment:
TODO: fix dropDisabled code that is overly friendly now that this is gated

Reply:
it has been fixed.